node/pkg/tss/comm/util.go

```diff
@@ -0,0 +1,193 @@
+package comm
+
+import (
+	"container/heap"
+	"context"
+	"crypto/x509"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	maxAttempts    = 10 // max backoff attempts before time doesn't increase.
+	minBackoffTime = time.Millisecond * 100
+	maxBackoffTime = minBackoffTime * 1024
+
+	connectionCheckTime = time.Second * 5
+)
+
+type dialWithBackoff struct {
+	hostname       string
+	attempt        uint64
+	nextRedialTime time.Time
+}
+
+// NOT THREAD SAFE! do NOT share between two different goroutines.
+type backoffHeap struct {
+	heap            []dialWithBackoff
+	timer           *time.Timer
+	alreadyInHeap   map[string]bool
+	attemptsPerPeer map[string]uint64 // on successful dial, reset to 0.
+}
+
+// Enqueue adds a hostname to the heap, with a new backoff time.
+func (d *backoffHeap) Enqueue(hostname string) {
+	if d.alreadyInHeap[hostname] {
+		return
+	}
+
+	if v, ok := d.attemptsPerPeer[hostname]; ok {
+		newv := v + 1
+		if newv >= maxAttempts {
+			newv = maxAttempts
+		}
+
+		d.attemptsPerPeer[hostname] = newv
+	} else {
+		d.attemptsPerPeer[hostname] = 0
+	}
+
+	elem := dialWithBackoff{
+		hostname: hostname,
+		attempt:  d.attemptsPerPeer[hostname],
+	}
+
+	elem.setBackoff()
+	heap.Push(d, elem)
+	d.alreadyInHeap[hostname] = true
+
+	d.setTopAsTimer()
+}
+
+func (d *backoffHeap) Dequeue() string {
+	if len(d.heap) == 0 {
+		return ""
+	}
+
+	elem, ok := heap.Pop(d).(dialWithBackoff)
+	if !ok {
+		return "" // shouldn't happen.
+	}
+
+	d.alreadyInHeap[elem.hostname] = false
+
+	d.setTopAsTimer()
+
+	return elem.hostname
+}
+
+func (d *backoffHeap) ResetAttempts(hostname string) {
+	delete(d.attemptsPerPeer, hostname)
+}
+
+func (d *backoffHeap) setTopAsTimer() {
+	if len(d.heap) == 0 {
+		d.stopAndDrainTimer() // no elements: stop the timer.
+
+		return
+	}
+
+	endTime := d.peek().nextRedialTime // we have at least one element.
+
+	d.stopAndDrainTimer()
+	d.timer.Reset(time.Until(endTime))
+}
+
+func (d *backoffHeap) stopAndDrainTimer() {
+	// stopping the timer, if its channel is not drained: drain it.
+	if !d.timer.Stop() && len(d.timer.C) > 0 {
+		select {
+		case <-d.timer.C:
```

Comment:
This is more of a precaution. It isn't needed necessarily, and it does work without it, but I thought it would be safer and would have no additional cost.